libminifi/include/utils/file/FileMatcher.h

```diff
@@ -0,0 +1,102 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <set>
+#include <utility>
+#include <memory>
+
+#include "utils/OptionalUtils.h"
+#include "core/logging/Logger.h"
+
+struct FileMatcherTestAccessor;
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+namespace file {
+
+class FileMatcher {
```

Comment:
Could this be a function? Its name sounds like an operation, like a `utils::file::match` function, that returns a list (vector) of paths.

Reply:
as the logic inside `FileMatcher` is pretty "involved", moving all of them into a single function would certainly cause more confusion, also we could think of `FileMatcher` as the equivalent of `std::regex`, as the pattern itself can contain errors (and `FileMatcher` logs them), this "error-state" is better represented construction time than usage time (here we could throw from the constructor, unclear if that is something we would like to do and stop loading the extension, currently it is on a "best-effort" basis)

should we load extensions even if we encounter an invalid pattern (and log errors) or should we refuse to load extension then? (the only problem with the latter is that then we won't load any extensions, not even the curl extension by which the c2-server could correct the pattern)

Reply:
The implementation could still involve breaking up the code to different classes, I would just prefer the interface to be a function, just like `std::regex_match` is one. I think FilePattern is a better analogy to `std::regex`, although it doesn't involve expensive regex compilation in this case. What do you think about `utils::file::match(std::vector<FilePattern>)`? `forEachFile` could become a loop at the usage side instead.

If the pattern is invalid, the FilePattern constructor could throw. If there are issues related to the filesystem, then the operation should throw. If a directory in a pattern doesn't exist, then that could be an empty match without failure, allowing us to load the rest of the extensions. There could be a function in ExtensionManager.cpp that converts the patterns string to a `vector<FilePattern>` and logs on error.

```
// ExtensionManager.cpp, anonymous namespace
std::vector<FilePattern> splitPatterns(std::string_view patterns, core::logging::Logger& logger) {
  std::vector<FilePattern> result;
  const auto patterns_vector = StringUtils::split(patterns, ",");
  result.reserve(patterns_vector.size());
  for (auto& pattern_str: patterns_vector) {
    auto pattern = FilePattern::fromPattern(pattern_str);  // I would consider making this function a constructor
    // currently this returns an optional, but if it threw, this would become a try block
    if (pattern) result.push_back(*pattern);
    else logger->log_warn("Invalid pattern: %s", pattern_str);
  }
  result.shrink_to_fit();
  return result;
}

// in ExtensionManager::initialize
    const auto files = utils::file::match(splitPatterns(*pattern, logger_));
```